src/Parsing/ParserState.php

```diff
@@ -487,13 +459,11 @@ private function strsplit($string)
     }
 
     /**
-     * @param string $haystack
-     * @param string $needle
-     * @param int $offset
+     * @param int<0, max> $offset
```

Comment:
`$offset` can be negative to indicate an offset from the end of the string.  (Note that the version of `substr` provided in this class does not support a negative offset, whereas this method does because it calls on to the built-in PHP functions.)